go/src/koding/kites/kloud/credential/mongo.go

```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+		}
+
+		if f.Provider != "" && cred.Provider != f.Provider {
+			// provider does not match, filter out
+			continue
+		}
+
+		c := &Cred{
+			Ident:    cred.Identifier,
+			Provider: cred.Provider,
+			Title:    cred.Title,
+			Perm: &MongoPerm{
+				Acc:   acc,
+				User:  user,
+				Cred:  cred,
+				Roles: UserRole,
+			},
+		}
+
+		db.log().Debug("fetched %+v", c)
+
+		if team, ok := teams[rel.SourceId]; ok {
+			c.Team = team.Slug
+			c.Perm.(*MongoPerm).Team = team
+		}
+
+		*creds = append(*creds, c)
+	}
+
+	return nil
+}
+
+func (db *mongoDatabase) SetCred(c *Cred) error {
+	f, ok := c.Perm.(*Filter)
+	if !ok {
+		return errors.New("invalid credential permission")
+	}
+
+	perm, err := db.Validate(f, c)
+	if err == nil {
+		c.Perm = perm
+		return nil
+	}
+
+	if !models.IsNotFound(err, "jCredential") {
+		return err
+	}
+
+	mPerm, ok := perm.(*MongoPerm)
+	if !ok {
+		return fmt.Errorf("unable to create credential: %s", err)
+	}
+
+	if mPerm.Acc == nil {
+		return fmt.Errorf("unable to create credential: missing %q account", f.User)
+	}
+
+	now := time.Now().UTC()
+
+	if c.Title == "" {
+		c.Title = mPerm.User.Name + " " + now.String()
+	}
+
+	mPerm.Cred = &models.Credential{
+		Id:          bson.NewObjectId(),
+		Provider:    c.Provider,
+		Identifier:  c.Ident,
+		Title:       c.Title,
+		OriginId:    mPerm.Acc.Id,
+		Verified:    false,
+		AccessLevel: "private",
+		Meta: &models.CredentialMeta{
+			CreatedAt:  now,
+			ModifiedAt: now,
+		},
+	}
+
+	if err := modelhelper.CreateCredential(mPerm.Cred); err != nil {
+		return err
+	}
+
+	rel := &models.Relationship{
+		Id:         bson.NewObjectId(),
+		TargetId:   mPerm.Cred.Id,
+		TargetName: "JCredential",
+		SourceId:   mPerm.Acc.Id,
+		SourceName: "JAccount",
+		As:         "owner",
+		TimeStamp:  now,
+	}
+
+	if err := modelhelper.AddRelationship(rel); err != nil {
+		return err
+	}
+
+	c.Perm = mPerm
+
+	return nil
+}
+
+func (db *mongoDatabase) Lock(cred *Cred) error {
+	err := db.MongoDB.Run(modelhelper.CredentialsColl, func(c *mgo.Collection) error {
+		change := mgo.Change{
+			Update: bson.M{
+				"$set": bson.M{
+					"assignee.inProgress": true,
+					"assignee.assignedAt": time.Now().UTC(),
+				},
+			},
+			ReturnNew: true,
+		}
+
+		_, err := c.Find(
+			bson.M{
+				"identifier":          cred.Ident,
+				"assignee.inProgress": bson.M{"$ne": true},
+			},
+		).Apply(change, &models.Credential{})
+
+		return err
+	})
+
+	if err == mgo.ErrNotFound {
+		return fmt.Errorf("credential %q is already locked", cred.Ident)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error locking %q credential: %s", cred.Ident, err)
+	}
+
+	return nil
+}
+
+func (db *mongoDatabase) Unlock(cred *Cred) error {
+	return db.MongoDB.Run(modelhelper.CredentialsColl, func(c *mgo.Collection) error {
+		return c.Update(
+			bson.M{"identifier": cred.Ident},
+			bson.M{"$set": bson.M{"assignee.inProgress": false}},
+		)
+	})
+}
+
+func match(existing, expected []string) bool {
```

Comment:
The `match` does not test for equality here, so we can't `len(existing) != len(expected)` here.

I've changed the func name to `contains`.